### ISSUE TITLE
Keepalived reload time issue

### ISSUE THREAD

**Issue by senthilnathann** [acassen/keepalived#1425]
Issue :-

During keepalived reload time ,Master went to Backup state ,after next second again Master become Master.Due to this GARP send to all VIP's during every keepalived reload time.

Why every keepalived reload time it's happens ?

Keepalived version:

Keepalived v2.0.7

OS : CentOS Linux release 7.4.1708 (Core)

Logs :: 

> Nov  7 16:07:25 l3-newversion-1 keepalived[16247]: Reloading keepalived: [  OK  ]
> Nov  7 16:07:25 l3-newversion-1 systemd[1]: Reloaded SYSV: Start and stop Keepalived.
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: Reloading
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: Opening file '/etc/keepalived/keepalived.conf'.
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: Assigned address 172.20.131.113 for interface bond0
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: Assigned address fe80::ec4:7aff:feb0:149e for interface bond0
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: (VI_GATEWAY) setting VIPs.
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: (VI_1) setting VIPs.
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: (VI_GATEWAY) removing VIPs.
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: (VI_GATEWAY) Entering BACKUP STATE (init)
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: (VI_1) removing VIPs.
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: (VI_1) Entering BACKUP STATE (init)
> Nov  7 16:07:25 l3-newversion-1 Keepalived_vrrp[15946]: VRRP sockpool: [ifindex(4), proto(112), unicast(0), fd(8,9)]
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: (VI_GATEWAY) Receive advertisement timeout
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: (VI_GATEWAY) Entering MASTER STATE
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: (VI_GATEWAY) setting VIPs.
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: (VI_GATEWAY) Sending/queueing gratuitous ARPs on bond0 for 172.20.131.115
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  7 16:07:26 l3-newversion-1 Keepalived_vrrp[15946]: Sending gratuitous ARP on bond0 for 172.20.131.115

**Comment by pqarmitage** [comment]
From my CentOS 7.7 system, the latest version of keepalived appears to be v1.3.5, so I am assuming that you are using a version of keepalived that you have built yourself. The output of `keepalived -v` would be helpful.

The first two lines of the log you show, I am not familiar with:
```
Nov 7 16:07:25 l3-newversion-1 keepalived[16247]: Reloading keepalived: [ OK ]
Nov 7 16:07:25 l3-newversion-1 systemd[1]: Reloaded SYSV: Start and stop Keepalived.
```
What process is logging with identifier `keepalived`. keepalived uses `Keepalived`, `Keepalived-vrrp` etc (i.e. initial leading capital letter), so the process logging `keepalived` is not keepalived :). Further, the PID of the process logging `keepalived` is not right for it to be the parent process of the keepalived processes running, since it should be only one or two lower than the PID of the keepalived vrrp process.

When I run `systemctl reload keepalived` I get the following output:
```
Nov 07 18:12:00 centos7 systemd[1]: Reloading LVS and VRRP High Availability Monitor.
Nov 07 18:12:00 centos7 Keepalived[3983]: Opening file '/etc/keepalived/keepalived.conf'.
Nov 07 18:12:00 centos7 systemd[1]: Reloaded LVS and VRRP High Availability Monitor.
Nov 07 18:12:00 centos7 Keepalived_vrrp[3985]: Registering Kernel netlink reflector
Nov 07 18:12:00 centos7 Keepalived_vrrp[3985]: Registering Kernel netlink command channel
Nov 07 18:12:00 centos7 Keepalived_vrrp[3985]: Registering gratuitous ARP shared channel
Nov 07 18:12:00 centos7 Keepalived_vrrp[3985]: Opening file '/etc/keepalived/keepalived.conf'.
Nov 07 18:12:00 centos7 Keepalived_vrrp[3985]: Truncating auth_pass to 8 characters
Nov 07 18:12:00 centos7 Keepalived_vrrp[3985]: VRRP_Instance(VI_2) setting protocol VIPs.
Nov 07 18:12:00 centos7 Keepalived_vrrp[3985]: Using LinkWatch kernel netlink reflector...
Nov 07 18:12:00 centos7 Keepalived_vrrp[3985]: VRRP sockpool: [ifindex(2), proto(112), unicast(0), fd(10,11)]
Nov 07 18:12:00 centos7 Keepalived_healthcheckers[3984]: Got SIGHUP, reloading checker configuration
Nov 07 18:12:00 centos7 Keepalived_healthcheckers[3984]: Initializing ipvs
Nov 07 18:12:00 centos7 Keepalived_healthcheckers[3984]: Opening file '/etc/keepalived/keepalived.conf'.
Nov 07 18:12:00 centos7 polkitd[758]: Unregistered Authentication Agent for unix-process:3989:556989 (system bus name :1.39, object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale en_GB.UTF-8) (disconnected from bus)
Nov 07 18:12:02 centos7 Keepalived_vrrp[3985]: VRRP_Instance(VI_2) Transition to MASTER STATE
```
which is quite different from what you are getting (I am not worried about you having no Keepalvied_healthcheckers entries).

Without knowing more about your configuration, how keepalived is being reloaded (including copies of the scripts that are being invoked) and a copy of your keepalived configuration it isn't going to be possible to help.

**Comment by senthilnathann** [comment]
I have using compiled keepalived version ,

My keepalived version is v2.0.7 .

> Keepalived v2.0.7 (08/23,2018)
> 
> Copyright(C) 2001-2018 Alexandre Cassen, <acassen@gmail.com>
> 
> Built with kernel headers for Linux 3.10.0
> Running on Linux 3.10.0-693.11.6.el7.x86_64 #1 SMP Thu Jan 4 01:06:37 UTC 2018
> 
> configure options: --prefix=/root/sanjay/keepalived/build/2.0.7/
> 
> Config options:  LVS VRRP VRRP_AUTH OLD_CHKSUM_COMPAT FIB_ROUTING
> 
> System options:  PIPE2 SIGNALFD INOTIFY_INIT1 VSYSLOG LIBNL3 RTA_ENCAP RTA_EXPIRES FRA_TUN_ID RTAX_CC_ALGO RTAX_QUICKACK FRA_OIFNAME IFA_FLAGS IP_MULTICAST_ALL NET_LINUX_IF_H_COLLISION LIBIPTC_LINUX_NET_IF_H_COLLISION LIBIPVS_NETLINK VRRP_VMAC SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE INET6_ADDR_GEN_MODE SO_MARK SCHED_RT SCHED_RESET_ON_FORK

Previously have used sysv init.d scripts to reload the keepalived , so we got small difference in logs.Now have tried to reload the keepalived using systemctl keepalived.service files,pls find the below logs,

Output of systemctl reload keepalived 

> systemd[1]: Reloaded LVS and VRRP High Availability Monitor.
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: Reloading
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: Opening file '/etc/keepalived/keepalived.conf'.
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: Assigned address 172.20.131.114 for interface bond0
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: Assigned address fe80::ec4:7aff:fe81:d36a for interface bond0
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: (VI_GATEWAY) setting VIPs.
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: (VI_1) setting VIPs.
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: (VI_GATEWAY) removing VIPs.
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: (VI_GATEWAY) Entering BACKUP STATE (init)
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: (VI_1) removing VIPs.
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: (VI_1) Entering BACKUP STATE (init)
> Nov  8 15:09:54 l3-newversion-2 Keepalived_vrrp[10716]: VRRP sockpool: [ifindex(4), proto(112), unicast(0), fd(8,9)]
> Nov  8 15:09:55 l3-newversion-2 Keepalived_vrrp[10716]: (VI_GATEWAY) Receive advertisement timeout
> Nov  8 15:09:55 l3-newversion-2 Keepalived_vrrp[10716]: (VI_GATEWAY) Entering MASTER STATE
> Nov  8 15:09:55 l3-newversion-2 Keepalived_vrrp[10716]: (VI_GATEWAY) setting VIPs.
> Nov  8 15:09:55 l3-newversion-2 Keepalived_vrrp[10716]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  8 15:09:55 l3-newversion-2 Keepalived_vrrp[10716]: (VI_GATEWAY) Sending/queueing gratuitous ARPs on bond0 for 172.20.131.115
> Nov  8 15:09:55 l3-newversion-2 Keepalived_vrrp[10716]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  8 15:09:55 l3-newversion-2 Keepalived_vrrp[10716]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  8 15:09:55 l3-newversion-2 Keepalived_vrrp[10716]: Sending gratuitous ARP on bond0 for 172.20.131.115
> 

Hope above logs will help you. 

My question is, Why keepalived went Backup state during keepalived reload time ?

**Comment by pqarmitage** [comment]
@senthilnathann Could you please provide a copy of your keepalived configuration, and also all keepalived and systemd logs from the time of running `systemctl start keepalived` through to after `systemctl reload keepalived`.

Do you have another keepalived instance so you are running a MASTER/BACKUP pair?

So far I have been unable to reproduce the problem, using CentOS 7.7 and keepalived v2.0.7.

**Comment by senthilnathann** [comment]
Yes , I have running MASTER/BACKUP mode keepalived. 

My OS version is CentOS Linux release 7.4.1708 (Core).

My Configuration file :  

> global_defs {
>         notification_email {
>                 XXXXX
>         }
>         notification_email_from XXXXXX
>         smtp_server 192.168.100.16
>         smtp_connect_timeout 30
>         router_id LVSID01
> }
> vrrp_sync_group VG1 {
>         group {
>                 VI_1
>                 VI_GATEWAY
>         }
> }
> vrrp_instance VI_GATEWAY {
>         interface bond0
>         virtual_router_id 91
>         priority 100
>         advert_int 1
>         smtp_alert
>         authentication {
>                 auth_type XXXXX
>                 auth_pass XXXXX
>         }
>         virtual_ipaddress {
>                 172.20.131.115
>         }
> }
> vrrp_instance VI_1 {
>         interface bond0
>         virtual_router_id 90
>         priority 100
>         advert_int 1
>         smtp_alert
>         authentication {
>                 auth_type XXXXX
>                 auth_pass XXXXX
>         }
> virtual_ipaddress {
>                 172.20.31.90
>                 172.20.31.92
>         }
> }
> virtual_server 172.20.31.90 80 {
>         delay_loop 10
>         lb_algo sh
>         lb_kind DR
> 
>         persistence_timeout 0
>         protocol TCP
>         real_server 172.20.16.67 80 {
>                 weight 1
>                 uthreshold 0
>                 lthreshold 0
> 
>                 TCP_CHECK {
>                         connect_timeout 3
>                         connect_port 80
>                 }
>         }

                                  

**systemctl start keepalived  time logs**

> Nov  8 16:04:08 l3-newversion-1 systemd[1]: Starting LVS and VRRP High Availability Monitor...
> Nov  8 16:04:08 l3-newversion-1 Keepalived[24893]: Starting Keepalived v2.0.7 (08/23,2018)
> Nov  8 16:04:08 l3-newversion-1 Keepalived[24893]: Running on Linux 3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 (built for Linux 3.10.0)
> Nov  8 16:04:08 l3-newversion-1 Keepalived[24893]: Command line: '/usr/sbin/keepalived' '-D'
> Nov  8 16:04:08 l3-newversion-1 Keepalived[24893]: Opening file '/etc/keepalived/keepalived.conf'.
> Nov  8 16:04:08 l3-newversion-1 systemd[1]: Started LVS and VRRP High Availability Monitor.
> Nov  8 16:04:08 l3-newversion-1 Keepalived[24894]: Starting Healthcheck child process, pid=24895
> Nov  8 16:04:08 l3-newversion-1 Keepalived[24894]: Starting VRRP child process, pid=24896
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: Registering Kernel netlink reflector
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: Registering Kernel netlink command channel
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: Opening file '/etc/keepalived/keepalived.conf'.
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: Assigned address 172.20.131.113 for interface bond0
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: Assigned address fe80::ec4:7aff:feb0:149e for interface bond0
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: Registering gratuitous ARP shared channel
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) removing VIPs.
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) removing VIPs.
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) Entering BACKUP STATE (init)
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Entering BACKUP STATE (init)
> Nov  8 16:04:08 l3-newversion-1 Keepalived_vrrp[24896]: VRRP sockpool: [ifindex(4), proto(112), unicast(0), fd(8,9)]
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) Receive advertisement timeout
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Receive advertisement timeout
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Entering MASTER STATE
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) setting VIPs.
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Sending/queueing gratuitous ARPs on bond0 for 172.20.31.90
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.92
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Sending/queueing gratuitous ARPs on bond0 for 172.20.31.92
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.92
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.92
> Nov  8 16:04:12 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.90
> 

systemctl reload keepalived time logs : 

> Nov  8 16:11:26 l3-newversion-1 systemd[1]: Reloaded LVS and VRRP High Availability Monitor.
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: Reloading
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: Opening file '/etc/keepalived/keepalived.conf'.
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: Assigned address 172.20.131.113 for interface bond0
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: Assigned address fe80::ec4:7aff:feb0:149e for interface bond0
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) setting VIPs.
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) setting VIPs.
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) removing VIPs.
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) Entering BACKUP STATE (init)
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) removing VIPs.
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Entering BACKUP STATE (init)
> Nov  8 16:11:26 l3-newversion-1 Keepalived_vrrp[24896]: VRRP sockpool: [ifindex(4), proto(112), unicast(0), fd(8,9)]
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) Receive advertisement timeout
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) Entering MASTER STATE
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) setting VIPs.
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: (VI_GATEWAY) Sending/queueing gratuitous ARPs on bond0 for 172.20.131.115
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: VRRP_Group(VG1) Syncing instances to MASTER state
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Entering MASTER STATE
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) setting VIPs.
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Sending/queueing gratuitous ARPs on bond0 for 172.20.31.90
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.92
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: (VI_1) Sending/queueing gratuitous ARPs on bond0 for 172.20.31.92
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.92
> Nov  8 16:11:27 l3-newversion-1 Keepalived_vrrp[24896]: Sending gratuitous ARP on bond0 for 172.20.31.90
>

**Comment by pqarmitage** [comment]
The reason for the instances going to backup state when reloading is that they are configured in a sync group (by this I mean that if the sync group is removed the vrrp instances do not go to backup state on reload).

From looking at the ChangeLog, commit 3ae1928 stopped vrrp instances in sync groups going to backup state on reload; that commit was first included in keepalived v2.0.8.

If you are upgrading keepalived I would recommend upgrading to v2.0.19, or better still the current head of the master branch which includes one or two further fixes, in particular commit 1c3c8ee relating to reloading. There were a significant number of commits relating to reloading between v2.0.10 and 2.0.19 (see the ChangeLog for details).
